Feat(anta_runner): Report should not report a fail when no tests are run

vitthalmagadum opened the pull request:
## Change Summary

Report should not report a fail when no tests are run

## Related Issue(s)

Fixes #5940 

## Component(s) name

`arista.avd.anta_runner`

## Proposed changes

Report should not report a fail when no tests are run

## How to test
Run molecule

## Checklist

### User Checklist
- N/A

### Repository Checklist
- [X] My code has been rebased from devel before I start
- [X] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [X] My change requires a change to the documentation and documentation have been updated accordingly.
- [X] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

vitthalmagadum committed:
Feat(anta_runner): Report should not report a fail when no tests are run

carl-baillargeon commented on ansible_collections/arista/avd/plugins/action/anta_workflow.py:
```diff
@@ -260,30 +260,32 @@ def build_reports(batch_results: Iterator[ResultManager], report_settings: dict[
-    if hide_statuses:
-        result_manager = result_manager.filter(hide=set(hide_statuses))
-
+    hide_statuses = hide_statuses if hide_statuses else []
+    filtered_result_manager = result_manager.filter(hide=set(hide_statuses))
```
```suggestion
    filtered_result_manager = result_manager.filter(hide=set(hide_statuses)) if hide_statuses else result_manager
```

vitthalmagadum replied:
Updated. Thanks!!

vitthalmagadum committed:
Merge branch 'devel' into anta_runner/reporting

carl-baillargeon commented on ansible_collections/arista/avd/molecule/anta_runner/converge.yml:
```diff
@@ -76,3 +76,15 @@
         anta_runner_tags: ["border_leaf"]
       run_once: true
+
+    - name: ANTA Filtered Run with hide statuses
```
```suggestion
    - name: ANTA Default Run with Filtered Report
```

carl-baillargeon replied:
Also, please move this after `ANTA Default Run` and use the same vars. Thanks.

carl-baillargeon commented on ansible_collections/arista/avd/molecule/anta_runner/converge.yml:
```diff
@@ -76,3 +76,15 @@
+        avd_catalogs_dir: "{{ root_dir }}/anta/avd_catalogs/filtered_run_hide"
+        anta_reports_dir: "{{ root_dir }}/anta/reports/filtered_run_hide"
+        anta_report_hide_statuses: ["unset"]
+        anta_runner_tags: ["border_leaf"]
```
```suggestion
        avd_catalogs_dir: "{{ root_dir }}/anta/avd_catalogs/default_run_filtered_report"
        anta_reports_dir: "{{ root_dir }}/anta/reports/default_run_filtered_report"
        anta_report_hide_statuses: ["unset"]
```

vitthalmagadum replied:
As discussed removed the molecule check.

carl-baillargeon commented on ansible_collections/arista/avd/plugins/action/anta_workflow.py:
```diff
@@ -260,30 +260,32 @@ def build_reports(batch_results: Iterator[ResultManager], report_settings: dict[
+    filtered_result_manager = result_manager.filter(hide=set(hide_statuses))
     # Sort the result manager
-    result_manager.sort(sort_by=["name", "categories", "test", "description", "result", "custom_field"])
+    filtered_result_manager.sort(sort_by=["name", "categories", "test", "description", "result", "custom_field"])
```
Let's add a warning log if the filtered manager is empty to say that the reports will be empty.

vitthalmagadum replied:
Added warning for empty reports.

carl-baillargeon commented on ansible_collections/arista/avd/plugins/action/anta_workflow.py:
```diff
@@ -260,30 +260,32 @@ def build_reports(batch_results: Iterator[ResultManager], report_settings: dict[
         path = Path(md_output_path)
-        md_report = MDReportGenerator()
-        md_report.generate(result_manager, path)
+        sections = [
+            (section, filtered_result_manager) if section.__name__ == "TestResults" else (section, result_manager)
+            for section in MDReportGenerator.DEFAULT_SECTIONS
+        ]
+        MDReportGenerator.generate_sections(md_filename=path, sections=sections)
```
Let's keep the existing `generate` method using the filtered manager (it could be empty but a warning is emitted now so all good).

vitthalmagadum replied:
Updated.
Thanks!

vitthalmagadum committed:
Addressed review comments: added warning for empty results

carl-baillargeon commented on ansible_collections/arista/avd/plugins/action/anta_workflow.py:
```diff
@@ -260,30 +260,31 @@ def build_reports(batch_results: Iterator[ResultManager], report_settings: dict[
+    filtered_result_manager.sort(sort_by=["name", "categories", "test", "description", "result", "custom_field"])
+
+    if not filtered_result_manager.results:
+        LOGGER.warning("No tests run/Reports are the empty due to filtered run on the user-defined ANTA catalogs")
```
The warning message looks weird. Can you rephrase it a little bit better please?

Also, this is not 100% true since `filtered_result_manager` can point to the original `result_manager` object if there is no filter.

On line 263 above, we could do something like:
```
if hide_statuses:
    filtered_result_manager = result_manager
else:
    filtered_result_manager = result_manager.filter(hide=set(hide_statuses))
    if len(filtered_result_manager) == 0:
        LOGGER.warning("...")
```

vitthalmagadum replied:
Updated. Thanks!!

carl-baillargeon committed:
Merge branch 'devel' into anta_runner/reporting

vitthalmagadum committed:
Merge branch 'devel' into anta_runner/reporting

vitthalmagadum committed:
updated logger messages

carl-baillargeon commented on ansible_collections/arista/avd/plugins/action/anta_workflow.py:
```diff
@@ -261,29 +261,34 @@ def build_reports(batch_results: Iterator[ResultManager], report_settings: dict[
-        result_manager = result_manager.filter(hide=set(hide_statuses))
+        filtered_result_manager = result_manager.filter(hide=set(hide_statuses))
+        if not filtered_result_manager.results:
+            msg = f"The reports are empty after filtering by hide_statuses. Hide Status(s): {', '.join(hide_statuses)}"
```
```suggestion
            msg = f"The report is empty because all results were hidden by the provided status filters: {', '.join(hide_statuses)}"
```

carl-baillargeon committed:
Update ansible_collections/arista/avd/plugins/action/anta_workflow.py

carl-baillargeon committed:
Merge branch 'devel' into anta_runner/reporting

carl-baillargeon reviewed (approved):
LGTM if CI passes.

gmuloc reviewed (commented):
@vitthalmagadum could we try to add a test for this?

gmuloc reviewed (dismissed):
blocking merging for now untill we see if adding a test is possible

carl-baillargeon committed:
Merge branch 'devel' into anta_runner/reporting

vitthalmagadum commented:
> @vitthalmagadum could we try to add a test for this?
Can I try adding a unit test using pytest for this? I had added Molecule earlier for testing purpose but removed it later.
@carl-baillargeon @gmuloc

carl-baillargeon commented:
> > @vitthalmagadum could we try to add a test for this?
> > Can I try adding a unit test using pytest for this? I had added Molecule earlier for testing purpose but removed it later.
> > @carl-baillargeon @gmuloc

@vitthalmagadum You can add back the molecule then. Apologies for the back and forth :)

gmuloc committed:
Merge branch 'devel' into anta_runner/reporting

vitthalmagadum committed:
added molecule scenario

vitthalmagadum committed:
Merge branch 'devel' into anta_runner/reporting

vitthalmagadum commented:
> > > @vitthalmagadum could we try to add a test for this?
> > > Can I try adding a unit test using pytest for this? I had added Molecule earlier for testing purpose but removed it later.
> > > @carl-baillargeon @gmuloc
> 
> @vitthalmagadum You can add back the molecule then. Apologies for the back and forth :)

Added molecule.
Thanks!!

carl-baillargeon committed:
Merge branch 'devel' into anta_runner/reporting
